extensions/amp-ad-template/0.1/amp-ad-template.js

```diff
@@ -0,0 +1,111 @@
+import {getAmpdoc} from '../../../src/service';
+import {Bind} from '../../amp-bind/0.1/bind-impl';
+
+/** @typedef {{
+      templatePromise: Promise<string>,
+      access: number
+    }} */
+export let CachedTemplateDef;
+
+/** @private {!Object<string, string|boolean>} */
+const TEMPLATE_CORS_CONFIG = {
+  mode: 'cors',
+  method: 'GET',
+  // This should be cached across publisher domains, so don't append
+  // __amp_source_origin to the URL.
+  ampCors: false,
+  credentials: 'omit',
+};
+
+/** @const {string} */
+const TAG = 'amp-ad-template';
+
+export class AmpAdTemplate {
+
+  /**
+   * @param {!Window} win
+   * @param {function(string)=} opt_onRetrieve Function to be executed upon
+   *   the successful retrieval of a template. If one is not supplied, we will
+   *   default to a no-op.
+   */
+  constructor(win, opt_onRetrieve = unusedTemplate => {}) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {!Object<number, !CachedTemplateDef>} */
+    this.templateCache_ = {};
+
+    /** @private {function(string)} */
+    this.onRetrieve_ = opt_onRetrieve;
+  }
+
+  /**
+   * Fetch and parse template from AMP cache.  Result is stored in global in
+   * order to reduce overhead when template is used multiple times.
+   * @param {number} templateId
+   * @return {!CachedTemplateDef}
+   */
+  retrieveTemplate(templateId) {
+    // Retrieve template from AMP cache.
+    this.templateCache_[templateId] = this.templateCache_[templateId] ||
+    {
+      templatePromise: Services.xhrFor(this.win_)
+          .fetchText(getMode(this.win_).localDev ?
+            `http://ads.localhost:${this.win_.location.port}` +
+            `/a4a_template/adzerk/${templateId}` :
+            `${urls.cdn}/c/s/adzerk/${templateId}`,
+          TEMPLATE_CORS_CONFIG)
+          .then(response => response.text())
+          .then(template => {
+            this.onRetrieve_(template);
+            return template;
+          }),
+      access: Date.now(),
+    };
+    const cacheKeys = /**@type {!Array<number>}*/
+        (Object.keys(this.templateCache_));
+    if (cacheKeys.length > 5) {
+      dev().warn(TAG, 'Trimming template cache');
+      // Evict oldest entry to ensure memory usage is minimized.
+      cacheKeys.sort((a, b) =>
+        this.templateCache_[b].access - this.templateCache_[a].access);
+      delete this.templateCache_[cacheKeys[cacheKeys.length - 1]];
+    }
+    dev().assert(this.templateCache_[templateId]);
+    return this.templateCache_[templateId];
+  }
+
+  /**
+   * @param {!JsonObject} templateValues The values to macro in.
+   * @param {!Element} element Parent element containing template.
+   * @param {!Window=} opt_window
+   */
+  populateTemplate(templateValues, element, opt_window) {
+    const bind = new Bind(getAmpdoc(element), opt_window);
+    bind.setState(templateValues);
+  }
+}
+
+AMP.extension('amp-ad-template', '0.1', AMP => {
+  AMP.registerElement('amp-ad-template', AmpAdTemplate);
```

Comment:
this should not be an extension. it can be just another file in `amp-a4a/0.1/` folder for code sharing

Reply:
Done.